### PR TITLE
feat(rules): allow and/or in conditions

### PR DESCRIPTION
**Pull request by sileht**
This allows to use and and or in conditions except for some complicated
attributes.

Fixes MRGFY-117
Fixes MRGFY-470
Related MRGFY-16

Change-Id: I135062de71ed647e95fef9c9153e385bacc53119

### CODE REVIEW

**Comment by sileht** [comment]
This pull request is part of a stack:
* feat(rules): allow and/or in conditions ([#2691](https://github.com/Mergifyio/mergify-engine/pull/2691))
* feat: allow merged/closed/author/base/head in and/or conditions ([#2692](https://github.com/Mergifyio/mergify-engine/pull/2692))
* feat(conditions): allow checks condition in and/or ([#2693](https://github.com/Mergifyio/mergify-engine/pull/2693))

File: mergify_engine/context.py [hunk 1]
```diff
@@ -673,6 +673,7 @@ async def consolidated_reviews(
         # And only keep the last review for each user.
         comments: typing.Dict[github_types.GitHubLogin, github_types.GitHubReview] = {}
         approvals: typing.Dict[github_types.GitHubLogin, github_types.GitHubReview] = {}
+
```
Comment (jd) [inline mergify_engine/context.py:676]: remove?

File: mergify_engine/rules/__init__.py [hunk 2]
```diff
@@ -75,55 +91,150 @@ async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
...
+                # TODO(sileht): remove this heuristics to allow this attribute everywhere
+                if attribute.partition("-")[0] in [
+                    "check",  # heuristic in merge/queue
+                    "status",  # heuristic in merge/queue
+                    "head",  # heuristic for hiding useless rules
+                    "base",  # heuristic for hiding useless rules
+                    "author",  # heuristic for hiding useless rules
+                    "merged_by",  # heuristic for hiding useless rules
+                    "merged",  # heuristic for detecting if pull request is merged by us
+                    "closed",  # heuristic for detecting if pull request is merged by us
```
Comment (jd) [inline mergify_engine/rules/__init__.py:145]: This is pretty bad; do you have a plan to remove this limitation in the end?

File: mergify_engine/rules/__init__.py [hunk 3]
```diff
@@ -75,55 +91,150 @@ async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
...
-    def __len__(self) -> int:
-        return len(self.conditions)
+    def walk(self) -> typing.Iterator[RuleCondition]:
+        for condition in self._walk(self.conditions):
+            yield condition
 
-    def copy(self) -> "RuleConditions":
-        return RuleConditions([c.copy() for c in self.conditions])
+    def is_faulty(self) -> bool:
+        return bool(first.first(c.evaluation_error for c in self.walk()))
```
Comment (jd) [inline mergify_engine/rules/__init__.py:185]:
```suggestion
        return any(c.evaluation_error for c in self.walk())
```

File: mergify_engine/rules/__init__.py [hunk 4]
```diff
@@ -75,55 +91,150 @@ async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
...
+        return textwrap.indent(summary, "  " * level)
+
+    def get_summary(self) -> str:
+        return self._walk_for_summary(self.conditions)
+
+
+def RuleConditions(
+    conditions: typing.List[typing.Union["RuleConditionGroup", RuleCondition]]
+) -> RuleConditionGroup:
+    return RuleConditionGroup({"and": conditions}, limit_leaf_types=False)
```
Comment (jd) [inline mergify_engine/rules/__init__.py:232]: 👍

File: mergify_engine/rules/__init__.py [hunk 5]
```diff
@@ -278,29 +389,41 @@ async def create(
...
+
+            # NOTE(sileht):
+            # In the summary, we display rules in four groups:
+            # * rules where all attributes match -> matching_rules
+            # * rules where only BASE_ATTRIBUTES match (filter out rule written for other branches) -> matching_rules
+            # * rules that won't work due to a configuration issue detected at runtime (eg team doesn't exists) -> faulty_rules
+            # * rules where only BASE_ATTRIBUTES don't match (mainly to hide rule written for other branches) -> ignored_rules
+            categorized = False
+
+            # FIXME(sileht): We may need to revisit this boolean
```
Comment (jd) [inline mergify_engine/rules/__init__.py:410]: can you write why?

File: mergify_engine/rules/__init__.py [hunk 6]
```diff
@@ -454,6 +577,26 @@ def YAML(v: bytes) -> typing.Any:
         raise YAMLInvalid(message="Invalid YAML", error_message=error_message)
 
 
+def RuleConditionSchema(v: typing.Any, depth: int = 0) -> typing.Any:
+    if depth > 3:
+        raise voluptuous.Invalid("maximun number of nested and/or reached")
```
Comment (jd) [inline mergify_engine/rules/__init__.py:582]: This does not seem documented.
```suggestion
    if depth > 3:
        raise voluptuous.Invalid("Maximum number of nested conditions reached")
```

File: mergify_engine/web/simulator.py [hunk 7]
```diff
@@ -166,17 +166,19 @@ async def simulator(
...
         status_code=200,
         content={
             "title": title,
             "summary": summary,
-            "conditions": {
-                "pull_request_rules": pull_request_rules_conditions,
-            },
+            # "conditions": {
+            #    "pull_request_rules": pull_request_rules_conditions,
+            # },
```
Comment (jd) [inline mergify_engine/web/simulator.py:182]: ?

File: mergify_engine/tests/functional/test_simulator.py [hunk 8]
```diff
@@ -122,7 +122,8 @@ async def test_simulator_with_token(self):
         assert r.json() == {
             "errors": [
                 "expected str @ pull_request_rules → item 0 → actions → label → remove → item 0",
-                "expected str @ pull_request_rules → item 0 → conditions → item 0",
+                "extra keys not allowed @ pull_request_rules → item 0 → conditions → item 0 → -conflict",
+                "required key not provided @ pull_request_rules → item 0 → conditions → item 0 → Any('or', 'and', msg=None)",
```
Comment (jd) [inline mergify_engine/tests/functional/test_simulator.py:126]: that last one is kinda weird, can we have a better string for the last element?

File: mergify_engine/rules/__init__.py [hunk 9]
```diff
@@ -75,55 +91,150 @@ async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
...
+                for _condition in cls._walk(condition.conditions):
+                    yield _condition
+            else:
+                raise RuntimeError(f"Unsupported condition type: {type(condition)}")
 
-    def __len__(self) -> int:
-        return len(self.conditions)
+    def walk(self) -> typing.Iterator[RuleCondition]:
+        for condition in self._walk(self.conditions):
+            yield condition
```
Comment (jd) [inline mergify_engine/rules/__init__.py:182]: I don't think you this at this point, you can rename `_walk` to `walk` AFAIU.

**Review by jd — DISMISSED** [review]

File: mergify_engine/rules/__init__.py (hunk 2 above)
Comment (sileht) [reply mergify_engine/rules/__init__.py:145]: I have found a solution for each one in further PR of the stack.

File: mergify_engine/rules/__init__.py (hunk 5 above)
Comment (sileht) [reply mergify_engine/rules/__init__.py:410]: I removed it, I fixed the issue in another PR

File: mergify_engine/rules/__init__.py [hunk 10]
```diff
@@ -75,55 +91,150 @@ async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
...
         self._used = True
-        self.match = await filter.Filter({"and": self.conditions})(obj)
+        self.match = await filter.Filter(
+            typing.cast(filter.TreeT, {self.operator: self.conditions})
+        )(obj)
         return self.match
 
-    def __getitem__(self, idx: int) -> RuleCondition:
-        return self.conditions[idx]
+    def iter_root_rule_conditions(
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/rules/__init__.py:160]: This is temporary function dropped in a couple of PR

File: mergify_engine/tests/functional/test_simulator.py (hunk 8 above)
Comment (sileht) [reply mergify_engine/tests/functional/test_simulator.py:126]: I found a way to only display the extra keys message, which is enough to understand that `-conflict:` is invalid.

**Review by jd — APPROVED** [review]

**Review by jd — DISMISSED** [review]
typo doc

**Review by jd — APPROVED** [review]

**Pull request merged**